Add QCOM Multiview Per View Viewport extension support

mazvalente opened the pull request:
Add support for the extension VK_QCOM_multiview_per_view_viewports which allows for multiple viewports to be submitted in a single view. RenderDoc already supports this viewport/scissor configuration, and this change will allow applications which submit multiple viewports and scissors using the extension to render correctly.

mazvalente committed:
Add QCOM Multiview Per View Viewport extension support

Add support for the extension VK_QCOM_multiview_per_view_viewports which allows for multiple viewports to be submitted in a single view. RenderDoc already supports this viewport/scissor configuration, and this change will allow applications which submit multiple viewports and scissors using the extension to render correctly.

baldurk commented on renderdoc/api/replay/vk_pipestate.h:
```diff
@@ -1114,6 +1114,13 @@ If the list is empty, fdm_offset is disabled and rendering is as normal.
+
+If only a single viewport is present in any view, multiviewPerViewViewports is set to false;
+)");
+  bool multiviewPerViewViewports;
```
This bool is added to the pipe state here but never actually filled in. Is this left over from an earlier draft, or is there more code to handle this that's missing?

mazvalente replied:
Leftover from an earlier draft, I'd thought I would need to add support there but it turns out it was unnecessary. Removed.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_device_funcs.cpp:
```diff
@@ -1917,6 +1917,15 @@ bool WrappedVulkan::Serialise_vkCreateDevice(SerialiserType &ser, VkPhysicalDevi
+      multiviewperviewviewports = true;
+      Extensions.push_back(VK_QCOM_MULTIVIEW_PER_VIEW_VIEWPORTS_EXTENSION_NAME);
+      RDCLOG("Enabling VK_QCOM_multiview_per_view_viewports");
+    }
```
Why are you forcibly enabling the extension here, and the feature below? I'm not sure I understand the reason for this, but especially for this extension where there's no more granular enable than the device feature this could actively break or change behaviour in captures that don't use the extension by changing which viewports are used.

baldurk commented on renderdoc/driver/vulkan/vk_overlay.cpp:
```diff
@@ -1347,7 +1347,7 @@ ResourceId VulkanReplay::RenderOverlay(ResourceId texid, FloatVector clearCol, D
           viewport.height = scissor.w;
 
-          vt->CmdSetViewport(Unwrap(cmd), 0, 1, &viewport);
+          vt->CmdSetViewport(Unwrap(cmd), 0, (uint32_t)state.views.size(), &viewport);
```
Both of these changes seem extremely broken as you're passing potentially a number of viewports to the function but only passing one stack variable, meaning it will read off the end and read garbage. What was this change intended to do?

baldurk commented on renderdoc/driver/vulkan/vk_rendertexture.cpp:
```diff
@@ -575,13 +575,8 @@ bool VulkanReplay::RenderTextureInternal(TextureDisplay cfg, const ImageState &i
-                           1.0f};
-    vt->CmdSetViewport(Unwrap(cmd), 0, 1, &viewport);
+    VulkanRenderState &state = m_pDriver->m_RenderState;
+    vt->CmdSetViewport(Unwrap(cmd), 0, (uint32_t)state.views.size(), state.views.data());
```
I don't understand what the purpose of this change is. The current viewports in the renderstate are not related to what texture rendering we're trying to do and may not make any sense. Why are you changing the viewport for this internal rendering here, and why only this one?

baldurk reviewed (changes requested):
The extension itself seems fairly straightforward but this pull request seems kind of unfinished - was it marked as ready for review accidentally and it should still be a draft?

mazvalente committed:
Clean up Commit and Uneccessary Changes

Removed remnants of earlier code commits that should not have been in the final PR and deleted an unneeded change which added a struct to the DeviceCreateInfo struct if it wasn't already present.

mazvalente commented:
> was it marked as ready for review accidentally and it should still be a draft?

It was marked as ready for review intentionally, but definitely should have received another pass before doing so, my apologies. I've cleaned up the suggested changes and will will reopen for review once I verify that everything is working properly.

baldurk commented on renderdoc/driver/vulkan/wrappers/vk_device_funcs.cpp:
```diff
@@ -4294,6 +4306,9 @@ VkResult WrappedVulkan::vkCreateDevice(VkPhysicalDevice physicalDevice,
     bufferAddressFeaturesEXT->bufferDeviceAddressCaptureReplay = VK_TRUE;
 
+  if(multiviewPerViewViewportsFeatures)
+    multiviewPerViewViewportsFeatures->multiviewPerViewViewports = VK_TRUE;
```
You're still enabling this feature even if the capture didn't use it. The same question applies as before - why are you forcibly enabling the extension, without any additional use on replay?

mazvalente replied:
I misunderstood the spec requirement, I thought that by including the struct in the pNexts you were automatically including the feature so I set the enablement. On rechecking khronos that is not the case for this extension, only the other extension I had been working with, so I'll remove this.

baldurk replied:
I don't believe that is how any extensions work. For the extensions that include a feature struct (which is most of them, I think only a small number of extensions are automatically enabled just by the extension string and have no feature struct) the feature must be enabled for them to work.

That isn't the primary issue here though - what you are doing is checking for the extension struct being passed in from the capture, and forcibly turning the feature on. If the struct was passed in then most likely the bool was already true so this is redundant. If the struct were passed in but the bool was false then you're now setting it to true and *changing the behaviour of the capture*. I don't see any reason to do that and it could break the replay of captures - albeit in a somewhat obtuse circumstance.

mazvalente replied:
I agree with you, this should not be present, and is removed in the latest update. As mentioned I had misunderstood the spec requirement, I should not have been changing the behavior here at all.

baldurk reviewed (changes requested):
I'm a little unsure of things - the only changes remaining in this PR are those which purely allow and serialise it (aside from one change which forcibly enables it which I've left a comment on). In the previous version though you had modifications to the replay side which seemed designed to update handling for it.

You haven't explained why the difference - does this extension really need no modification whatsoever of the replay handling? If so, why were there changes before and how have you tested this to make sure that the replay handling is robust even when the extension is used?

mazvalente commented:
> You haven't explained why the difference - does this extension really need no modification whatsoever of the replay handling? If so, why were there changes before and how have you tested this to make sure that the replay handling is robust even when the extension is used?

You're right, this does require some clarification. All that multiviewperviewviewports does is give permission for Vulkan to use multiple viewports and scissors, which is already supported by the current VkCmdSetViewports/VkCmdSetScissors functions. From my testing with an app which submits multiple viewports and scissors it looks like there is no change needed on the replay side to enable the feature. Your mentioning that the feature should not be automatically enabled if the struct is attached suggests that I should add a check when submitting viewports to ensure that we have a form of multiviewport support enabled and otherwise either only accept the first submission or throw an error.

baldurk commented:
Something sounds wrong here. Your description of what the extension does doesn't match up with its description in the spec or my understanding of it. Unextended vulkan absolutely does allow the use of multiple viewports and scissor regions and there is no extension needed for that, though there are extensions that expand its use like `VK_EXT_shader_viewport_index_layer`. This extension seems to change the behaviour of multiview passes to automatically use the multiview view index as a viewport index if no explicit viewport index is selected.

I don't understand what you're saying about only accepting 'the first submission' or throwing any kind of error. Can you explain exactly what you mean by that?

mazvalente committed:
Remove forced extension enablement

Changes code to no longer automatically enable the extension if the extension struct is present.

mazvalente commented:
You're right, I was misunderstanding the description. After testing the implementation by making shader changes to selectively enable multiview in shaders through VK_EXT_shader_viewport_index_layer, it appears that no replay changes are necessary. Without shader changes the multiview functionality follows the spec implementation of using the multiview index as viewport index, and if I set the viewportIndex in the shader I override that behavior instead following the VK_EXT_shader_viewport_index_layer logic.

> I don't understand what you're saying about only accepting 'the first submission' or throwing any kind of error. Can you explain exactly what you mean by that?

I did not properly understand what I should have been doing to handle non-spec compliant situations, and was making bad assumptions, both on what the correct spec requirements were, and how to handle them. Please ignore it.

I've also uncovered one issue that I'll need to fix before putting this back in review, the debug overlay for Viewports/Scissors does not recognize the second viewport properly and is currently not displaying any overlay when selected for slice 1. I'll reopen this when I have a fix.

mazvalente committed:
Add Overlay.ViewportScissor support for MVPVV

Modifies RenderOverlay to support multiple viewports and scissors and update the Viewport/Scissor overlay to render the correct scissor for the given view.

mazvalente committed:
Merge remote-tracking branch 'upstream/v1.x' into Implement_multiviewperview_viewports

mazvalente commented:
This resolves the issue with vk_overlay not displaying the proper viewport/scissors region. If you know of a better way to retrieve the active slice value without modifying the RenderOverlay signature please advise me, but I don't believe it's retrievable without either changing the Overlay value to not be the enum or adding a new value in.

mazvalente committed:
Add MVPVV gate on overlay slices

Add a dependency on multiviewperviewviewports being enabled to allow for the slice to determine the viewport/scissor allocation. This is not guaranteed to be the case if the feature is not enabled.

baldurk commented:
I'm not meaning to be rude but I think there's still a misunderstanding here about how this extension works, and maybe how extensions work on Vulkan in general.

For that reason I'm closing this PR for now - could you please open this as a feature request issue for extension support and then it can be tracked that way?
